### DISCUSSION TITLE
Cluster stuck in "Failing over" state

### DISCUSSION THREAD

**Discussion by davidpanic** [cloudnative-pg/cloudnative-pg#1308] [Q&A]
I've been messing around with cnpg to see if it is a viable solution for our environment and it seemed to work fine until I tried to drain a node. Now my cluster is stuck in a "Failing over" state and I'm not sure how to fix it.

This is the YAML that I'm using:

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  namespace: cnpg-clusters
  name: aaa
spec:
  instances: 3
  minSyncReplicas: 1
  maxSyncReplicas: 1
  primaryUpdateStrategy: unsupervised
  storage:
    size: 20Gi
```

Here is the current status of the cluster:

```
Cluster Summary
Name:               aaa
Namespace:          cnpg-clusters
System ID:          7185152831789264915
PostgreSQL Image:   ghcr.io/cloudnative-pg/postgresql:15.1
Primary instance:   aaa-1
Status:             Failing over Failing over from aaa-2 to aaa-1
Instances:          3
Ready instances:    1
Current Write LSN:  0/14000000 (Timeline: 5 - WAL File: 000000050000000000000013)

Certificates Status
Certificate Name  Expiration Date                Days Left Until Expiration
----------------  ---------------                --------------------------
aaa-ca            2023-04-05 12:57:11 +0000 UTC  89.18
aaa-replication   2023-04-05 12:57:11 +0000 UTC  89.18
aaa-server        2023-04-05 12:57:11 +0000 UTC  89.18

Continuous Backup status
Not configured

Streaming Replication status
Name   Sent LSN    Write LSN   Flush LSN   Replay LSN  Write Lag  Flush Lag  Replay Lag  State      Sync State  Sync Priority
----   --------    ---------   ---------   ----------  ---------  ---------  ----------  -----      ----------  -------------
aaa-2  0/14000000  0/14000000  0/14000000  0/14000000  00:00:00   00:00:00   00:00:00    streaming  async       0

Unmanaged Replication Slot Status
No unmanaged replication slots found

Instances status
Name   Database Size  Current LSN  Replication role  Status             QoS         Manager Version  Node
----   -------------  -----------  ----------------  ------             ---         ---------------  ----
aaa-1  29 MB          0/14000000   Primary           OK                 BestEffort  1.18.1           [redacted]3
aaa-2  -              -            -                 pod not available  BestEffort  -                [redacted]2
```

The status seems to indicate the node *is* streaming data, but the status and logs indicate that the pod is in a broken state.

The logs of aaa-2 seem to just be these two lines repeating, nothing useful to help me fix the issue.

```
{"level":"info","ts":1672994054.3080347,"msg":"Ignore minSyncReplicas to enforce self-healing","logging_pod":"aaa-2","syncReplicas":0,"minSyncReplicas":1,"maxSyncReplicas":1}
{"level":"info","ts":1672994055.7038379,"msg":"Instance status probe failing","logging_pod":"aaa-2","err":"failed to connect to `host=/controller/run user=postgres database=postgres`: server error (FATAL: the database system is not yet accepting connections (SQLSTATE 57P03))"}
```

Would you advise me on what I should do to fix my cluster? Is there something I'm missing here?

I should mention that I'm using longhorn for storage, but all of the volumes are indicated as "healthy" in the UI.

**Comment by davidpanic** [comment]
By deleting both the volumes and pods of aaa-2 and aaa-3 it seems to have created new ones named aaa-4 and aaa-5 which have joined the cluster correctly and the cluster is now in a healthy state.

In my opinion this kind of task shouldn't require this kind of rebuild or manual intervention.

**Comment by sxd** [comment]
Hi @davidpanic

 A couple of things happens here, and others that we don't know, for example, what kind of storage are you using? when draining the node the PVC were still there or were in a bad state? The operator will retry to use a PVC if is there, but if the PVC is there (for the k8s cluster) but it's not being provided (the storage provider fails) this can cause this kind of issues

On the other hand, since this it's a database and a storage thing (we're supposing that it's a pvc problem or storage problem) that will require a manual intervention all the time because you may not want to lose data, and also, the operator knows exactly what Kubernetes knows, that the PVC is there and it's waiting for it.

Did you check the events in your cluster and the describe of the pods that were having issues?

I'm turning this into a discussion since it doesn't look like an issue, yet

Regards

**Comment by davidpanic** [comment]
I'm using longhorn for storage, and I'm specifically forcing the volumes to "strict-local" because having them on different nodes would be quite dumb for performance. There aren't any errors reported for them from what I can tell. If you think this is a storage issue I'll try using rancher's local path provisioner or something and I'll report back if I find anything else.

**Comment by davidpanic** [comment]
Ah, it seems like this is an issue with storage after all, Longhorn is mounting the volumes as read only because it thinks they are broken but in reality they just need to be re-attached. Here's the related issue: https://github.com/longhorn/longhorn/issues/1500
